python/ray/data/_internal/execution/operators/actor_pool_map_operator.py

```diff
@@ -210,8 +211,17 @@ def _dispatch_tasks(self):
             ).remote(DataContext.get_current(), ctx, *input_blocks)
 
             def _task_done_callback(actor_to_return):
-                # Return the actor that was running the task to the pool.
-                self._actor_pool.return_actor(actor_to_return)
+                if actor_to_return in self._actor_pool._num_tasks_in_flight:
+                    # Return the actor that was running the task to the pool.
+                    self._actor_pool.return_actor(actor_to_return)
+                else:
+                    assert (
+                        actor_to_return.get_location
+                        in self._actor_pool._restarting_actors
+                    )
+                    # Move the actor from restarting to running state.
+                    self._actor_pool.restarting_to_running(actor_to_return)
```

Comment:
We need to return the actor in this case as well. Otherwise, the actor will be no longer usable.
BTW, let's add a unit test in `test_actor_pool_map_operator.py` to cover this case.